tao_compiler/mlir/disc/IR/hlo_disc_ops.cc

```diff
@@ -699,6 +699,62 @@ LogicalResult SparseSegmentMeanOp::verify() {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// WhereOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult WhereOp::reifyReturnTypeShapes(
+    OpBuilder& builder, ValueRange operands,
+    SmallVectorImpl<Value>& reifiedReturnShapes) {
+  WhereOp::Adaptor adaptor(operands);
+  Location loc = this->getLoc();
+  auto input_type = adaptor.input().getType().cast<RankedTensorType>();
+
+  SmallVector<Value, 2> index_shape_values, num_output_elements_shape_values;
+
+  Value idx_one = builder.create<arith::ConstantIndexOp>(loc, 1);
+  auto input_rank = input_type.getRank();
+  Value num_input_elements = idx_one;
+  for (int i = 0; i < input_rank; i++) {
+    Value dim_i = builder.create<tensor::DimOp>(loc, operands[0], i);
```

Comment:
Does it need an index cast here? Also, we prefer the bit-32 integer type in DISC. cc @wyzero

Reply:
it depends on the semantic of the op. The general rules are:

- use index type in `reifyReturnTypeShapes` if we know nothing about the range.
- use the same type as the shape operand of the op if we know it's safe.
- index type is ok since we can always lower it to i32 when lowering to llvm (because we know index type is always used for shape related calculation). i64 type is another story. it's not safe to    downcast it to i32 by default since i64 is not always used for shape.